### PR TITLE
fix(broker): fixed missing distributed CREATEs during restart

### LINKED ISSUE #2110: Success response for push deployment is send before it is persistently written to logstorage
[PushDeploymentRequestHandler](https://github.com/zeebe-io/zeebe/blob/23cdd720546b0bada9e95b73a2faffe1bd3f9bb4/broker-core/src/main/java/io/zeebe/broker/system/management/deployment/PushDeploymentRequestHandler.java#L111) responds with a success before the CREATE record is committed to the logstream. When the broker restarts after sending the response, but before writing persistently to the logstream, the deployment partition assumes that the previous distribution was successful and does not re-distribute. This is the reason that the tests in `CreateDeploymentMultiplePartitionTest is flaky` (See issue #1250)

### PR DESCRIPTION
**Pull request by miguelpires**
closes #2110

### CODE REVIEW

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Hey @MiguelPires ,

please avoid to add additional dependencies to the engine module, especially atomix if not really necessary.

You can create an interface for that, which accepts the deployment key. And the implementation, which lies in the broker creates the response object and sends to the corresponding topic via atomix. With that we do not need atomix in the engine nor the SBE management schema for the PushDeploymentResponse. If you need help in how to change that just ping me. This should also be easy to impl in the EngineRule then.

In the Distributor I would expect that the subscription is opened for one deployment distribution and closed after all deployments are distributed. But currently we open a subscription for each partition-distribution (like request - response).  I assume this is a bit inefficient.

Would also be nice if we could verify the behavior with a test.

I hope my objections are understandable if not we can discuss them and do not hesitate to ask questions!

**Comment by MiguelPires** [comment]
Thanks for the feedback @Zelldon. I changed the response to go through an interface like you mentioned (and reverted the dependency changes) and made some modifications to the DeploymentDistributor to support opening a single subscription per deployment. While doing this I noticed a possible bug in how the distributor currently works which, for documentation purposes, is [here](https://github.com/zeebe-io/zeebe/issues/2735)

File: broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java [hunk 1]
```diff
@@ -237,4 +299,15 @@ private void handlePushResponse() {
...
     }
   }
+
+  public static String getDeploymentResponseTopic(final long deploymentKey) {
+    return String.format("deployment-response-%d", deploymentKey);
+  }
+
+  private Int2IntHashMap getPartitionResponses(final String topic) {
+    final Int2IntHashMap responses =
+        deploymentResponses.getOrDefault(topic, new Int2IntHashMap(NO_RESPONSE));
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: It seems a bit weird with the `NO_RESPONSE` and `RESPONSE` constants, which could be btw also a enum.
Why not using the same logic where you have the list of partition ids and remove the id from the list from which you get the response. In the end the list should be empty and you now you can write the distributed event.

Comment (Zelldon) [reply broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: Actually I also not seeing where you really use this values at all?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @MiguelPires for updating much better just some small things

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: engine/src/main/java/io/zeebe/engine/processor/workflow/DeploymentResponder.java [hunk 2]
```diff
@@ -15,9 +15,9 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package io.zeebe.broker.services;
+package io.zeebe.engine.processor.workflow;
 
-public class DispatcherSubscriptionNames {
-  public static final String TRANSPORT_CONTROL_MESSAGE_HANDLER_SUBSCRIPTION =
-      "control-message-handler";
+public interface DeploymentResponder {
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/DeploymentResponder.java]: Hm, I'm not sure why that's happening. I removed that dead code in the first commit and add DeploymentResponder in the second. When I look at the second commit the diff is normal.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e6c60504db8e4b148961dbf6267a5aeccaebdea
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: engine/src/main/java/io/zeebe/engine/processor/workflow/DeploymentResponder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/DeploymentResponder.java]: Hm I don't know :man_shrugging:

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e6c60504db8e4b148961dbf6267a5aeccaebdea
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by MiguelPires** [comment]
Hey @Zelldon. I made a few adjustments with your suggestions

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e6c60504db8e4b148961dbf6267a5aeccaebdea
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java [hunk 3]
```diff
@@ -304,10 +294,13 @@ public static String getDeploymentResponseTopic(final long deploymentKey) {
     return String.format("deployment-response-%d", deploymentKey);
   }
 
-  private Int2IntHashMap getPartitionResponses(final String topic) {
-    final Int2IntHashMap responses =
-        deploymentResponses.getOrDefault(topic, new Int2IntHashMap(NO_RESPONSE));
-    deploymentResponses.put(topic, responses);
+  private IntArrayList getPartitionResponses(final String topic) {
+    IntArrayList responses = deploymentResponses.get(topic);
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: I think you can use `computeIfAbsent` for this :)

File: broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java [hunk 4]
```diff
@@ -184,12 +207,42 @@ private void pushDeploymentToPartition(final int partitionLeaderId, final int pa
                 }));
   }
 
-  private void handleResponse(byte[] response, int partitionLeaderId, int partition) {
+  private void createResponseSubscription(
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: We are re-creating the subscription on retry does this make sense? Is this problematic ?

**Review by ChrisKujawa — COMMENTED** [review]
Just some small things

File: broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java (hunk 4 above)
Comment (MiguelPires) [reply broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: I couldn't find any documentation regarding that so to be safe I added that check to only create a subscription if we don't find one with that topic (line 212).

Comment (Zelldon) [reply broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: Ah sorry have overseen that. Might make sense to just open it at 
`pushDeployment(
      final long key, final long position, final DirectBuffer buffer)` ?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e6c60504db8e4b148961dbf6267a5aeccaebdea
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java (hunk 4 above)
Comment (MiguelPires) [reply broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: There are some paths were the don't push the deployment (like if there is only one partition) and then we have this permanently open subscription I think. I'd have to add additional checks to make sure that didn't happen

Comment (Zelldon) [reply broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: Ok but on `deployOnMultiplePartitions` it will work?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e6c60504db8e4b148961dbf6267a5aeccaebdea
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java (hunk 4 above)
Comment (MiguelPires) [reply broker-core/src/main/java/io/zeebe/broker/engine/impl/DeploymentDistributorImpl.java]: Sure, I moved it now

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 21232a09bababc8c21db272682e6c42dbf73e259
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e6c60504db8e4b148961dbf6267a5aeccaebdea
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Review by ChrisKujawa — APPROVED** [review]

**Comment by MiguelPires** [comment]
bors retry

**Pull request merged**